### DISCUSSION TITLE
Linux版本的go-judge   time依旧有时候为0，请问runtime可以作为oj网站题目的代码的运行时间吗？

### DISCUSSION THREAD

**Discussion by xxdcz11** [criyle/go-judge#108] [Q&A] [answered]
1.Linux版本go-judge的time有大于二十分之一，但不到十分之一的可能性为0，一个不准确的推测是time小于1ms的时候，就统计为0                         2.Linux版本go-judge的runtime不会为0，统计了解决二百条数据，没有出现runtime为0的情况     
3.我开发了一个oj网站，题目需要有时间限制，那么就需要统计题目代码运行的时间，所以请问Linux版本go-judge的runtime可以作为题目代码运行的时间吗？因为time有时候统计为0，我考虑使用runtime作为oj网站的题目代码的运行时间，go-judge文档里所说的runtime为程序运行现实时间，我感觉是可以作为oj网站题目的代码的运行时间的  
 4.Linux版本go-judge的runtime，会有您昨天说的“此字段只能提供一个程序运行的观察时间，而不来源于内核统计数据，因此并不准确。”这种问题吗？

**Comment by criyle** [comment, marked as answer]
1. 如果不是以 root 用户运行的 `go-judge` 默认使用的是 `wait4` 系统调用返回的时间，精度为 `1ms`。需要精确统计，请使用 root 帐户运行以给予 `cgroup` 使用权限。
2. `runtime` 相当于在开始运行程序的时候同时开启一个秒表计时，然后在程序停止的时候结束计时。在高系统负载（例如同时多线程评测） 的情况下，并不能确认计时中 100% 的时间程序处于运行状态（例如被其他程序抢占）。
3. 您可以使用这个指标，只是要注意到可能的误差。
4. Linux, macOS 下的 `runtime` 统计口径一致。windows 由于疏忽尚未实现。

**Reply by xxdcz11** [reply]
感谢您的及时回答！感动至极！我用sudo go-judge命令开启go-judge后，统计了150条数据，没有出现time为0的情况了，看来我还是用time作为程序运行时间吧
